source/DoublyLinkedList.h

```diff
@@ -0,0 +1,274 @@
+#ifndef __DATA_STRUCTURES_DOUBLY_LINKED_LIST
+#define __DATA_STRUCTURES_DOUBLY_LINKED_LIST
+
+#include <functional>
+#include <iterator>
+#include <type_traits>
+
+namespace DataStructures {
+
+template<typename T>
+class DoublyLinkedList
+{
+public:
+  static_assert(std::is_same<decltype(T{} == T{}), bool>(),
+                "value type must have `operator==(T&)` defined");
+  static_assert(std::is_same<decltype(T{} != T{}), bool>(),
+                "value type must have `operator!=(T&)` defined");
+
+private:
+  struct Element
+  {
+    T datum;
+    Element* previous;
+    Element* next;
+  };
+
+  Element* first;
+  Element* last;
+  size_t number_of_elements;
+
+public:
+  /**
+    A type for iterating forward through the list.
+  */
+  class iterator : std::iterator<std::forward_iterator_tag, T>
+  {
+  public:
+  private:
+    Element* current;
+
+  public:
+    explicit iterator(Element* start)
+      : current(start)
+    {}
+    explicit iterator()
+      : current(nullptr)
+    {}
+    iterator& operator++();
+    bool operator==(iterator other) const;
+    bool operator!=(iterator other) const;
+    T operator*() const;
+  };
+
+  /**
+    A type for iterating backward through the list.
+  */
+  class reverse_iterator : std::iterator<std::forward_iterator_tag, T>
```

Comment:
Do you need both `iterator` and `reverse_iterator` types? (As opposed to one iterator type that can be fed a direction at creation time and then typedef'd to both fwd and reverse iterator types.)

Reply:
oh and then have like `using iterator = iterator_t<forward>; using reverse_iterator = iterator_t<reverse>;`